src/OrleansCodeGenerator/Utilities/SyntaxFactoryExtensions.cs

```diff
@@ -311,12 +313,21 @@ public static TypeParameterConstraintClauseSyntax[] GetTypeConstraintSyntax(this
                         parameterConstraints.Add(SyntaxFactory.ClassOrStructConstraint(SyntaxKind.StructConstraint));
                     }
 
+                    // Follow with the base class or interface constraints.
                     foreach (var genericType in genericParameter.GetGenericParameterConstraints())
                     {
+                        // If the "struct" constraint was specified, skip the corresponding "ValueType" constraint.
+                        if (genericType == typeof(ValueType))
+                        {
+                            continue;
+                        }
+
                         parameterConstraints.Add(SyntaxFactory.TypeConstraint(genericType.GetTypeSyntax()));
                     }
 
-                    if (attributes.HasFlag(GenericParameterAttributes.DefaultConstructorConstraint))
+                    // The "new()" constraint must be the last constraint in the sequence.
+                    if (attributes.HasFlag(GenericParameterAttributes.DefaultConstructorConstraint)
+                        && !attributes.HasFlag(GenericParameterAttributes.NotNullableValueTypeConstraint))
```

Comment:
Here we only emit the `new()` constraint if the `struct` constraint hasn't been emitted.